File: backend/app/models/analysis.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


# --- Enums ---

class Jurisdiction(str, Enum):
    CHARLESTON = "charleston"
    MOUNT_PLEASANT = "mount_pleasant"
    NORTH_CHARLESTON = "north_charleston"
    SULLIVANS_ISLAND = "sullivans_island"
    ISLE_OF_PALMS = "isle_of_palms"
    FOLLY_BEACH = "folly_beach"
    JAMES_ISLAND = "james_island"
    KIAWAH = "kiawah"
    SUMMERVILLE = "summerville"
    GOOSE_CREEK = "goose_creek"
    HANAHAN = "hanahan"
    UNINCORPORATED = "unincorporated"


class UseType(str, Enum):
    RESIDENTIAL = "residential"
    COMMERCIAL = "commercial"
    MIXED_USE = "mixed_use"
    HOSPITALITY = "hospitality"
    ADAPTIVE_REUSE = "adaptive_reuse"


class RiskLevel(str, Enum):
    LOW = "Low"
    MODERATE = "Moderate"
    HIGH = "High"


# --- Request ---

class AnalysisRequest(BaseModel):
    jurisdiction: Jurisdiction
    address: str
    zoning_district: Optional[str] = None
    height_district: Optional[str] = None
    on_peninsula: Optional[bool] = None
    flood_zone: Optional[str] = None
    historic_overlay: Optional[str] = None
    lot_size_sf: Optional[float] = None
    use_types: Optional[List[UseType]] = None
    approximate_scale: Optional[str] = None
    existing_conditions: Optional[str] = None
    additional_context: Optional[str] = None


# --- Response sub-models ---

class ParcelInfo(BaseModel):
    address: str
    jurisdiction: str
    jurisdiction_display: str
    zoning_district: Optional[str] = None
    lot_size_sf: Optional[float] = None


class Setbacks(BaseModel):
    front: Optional[float] = None
    side: Optional[float] = None
    rear: Optional[float] = None
    front_rear_total: Optional[float] = None
    side_sw: Optional[float] = None
    side_ne: Optional[float] = None
    setback_notes: Optional[str] = None


class DevelopmentEnvelope(BaseModel):
    zoning_district: str
    zoning_description: str
    permitted_uses: List[str]
    conditional_uses: List[str]
    max_height_ft: Optional[float] = None
    max_stories: Optional[int] = None
    far: Optional[float] = None
    max_lot_coverage_pct: Optional[float] = None
    density_units_per_acre: Optional[float] = None
    setbacks: Setbacks
    buildable_area_sf: Optional[float] = None
    parking_requirements: Optional[str] = None
    lot_occupancy_pct: Optional[float] = None
    height_source: Optional[str] = None
    binding_constraint: Optional[str] = None


class Scenario(BaseModel):
    name: str
    unit_count_range: Optional[str] = None
    description: str
    constraints: List[str]
    risk_level: RiskLevel
    estimated_timeline: str
    board_engagement: str


class RiskMap(BaseModel):
    historic_overlay: Optional[str] = None
    flood_zone: Optional[str] = None
    building_category: Optional[str] = None
    accommodation_overlay: Optional[str] = None
    community_sensitivity: Optional[str] = None
    recent_board_decisions: Optional[str] = None
    environmental_constraints: Optional[str] = None


class ProcessTimeline(BaseModel):
    required_boards: List[str]
    review_sequence: str
    estimated_meetings: str
    estimated_months: str
    current_backlog: Optional[str] = None
    additional_permits: List[str]


class CostFraming(BaseModel):
    permit_fees_estimate: Optional[str] = None
    construction_cost_range: Optional[str] = None
    tax_credit_eligibility: Optional[str] = None
    bailey_bill_eligible: Optional[bool] = None
    total_cost_range: Optional[str] = None
    construction_type: Optional[str] = None
    base_hard_cost_range: Optional[str] = None
    applicable_premiums: Optional[List[str]] = None
    premium_adjusted_range: Optional[str] = None
    all_in_estimate_range: Optional[str] = None
    impact_fees_estimate: Optional[str] = None
    cws_fee_warning: Optional[str] = None


class ScenarioCommentary(BaseModel):
    by_right: str = ""
    optimized: str = ""
    with_variance: str = ""


class AnalysisMetadata(BaseModel):
    generated_at: datetime
    solver_version: str
    ai_model: str
    jurisdiction_module_version: str


# --- Response ---

class AnalysisResponse(BaseModel):
    parcel: ParcelInfo
    envelope: DevelopmentEnvelope
    scenarios: List[Scenario] = Field(..., min_length=3, max_length=3)
    risk_map: RiskMap
    process_timeline: ProcessTimeline
    cost_framing: CostFraming
    metadata: AnalysisMetadata
    confidence_tier: int
    confidence_label: str
    disclaimer: Optional[str] = None

File: frontend/src/types/analysis.ts
// --- Enums ---

export type Jurisdiction =
  | "charleston"
  | "mount_pleasant"
  | "north_charleston"
  | "sullivans_island"
  | "isle_of_palms"
  | "folly_beach"
  | "james_island"
  | "kiawah"
  | "summerville"
  | "goose_creek"
  | "hanahan"
  | "unincorporated";

export type UseType =
  | "residential"
  | "commercial"
  | "mixed_use"
  | "hospitality"
  | "adaptive_reuse";

export type RiskLevel = "Low" | "Moderate" | "High";

// --- Request ---

export interface AnalysisRequest {
  jurisdiction: Jurisdiction;
  address: string;
  lot_size_sf?: number;
  use_types?: UseType[];
  approximate_scale?: string;
  existing_conditions?: string;
  additional_context?: string;
}

// --- Response sub-models ---

export interface ParcelInfo {
  address: string;
  jurisdiction: string;
  jurisdiction_display: string;
  zoning_district?: string;
  lot_size_sf?: number;
}

export interface Setbacks {
  front?: number;
  side?: number;
  rear?: number;
  front_rear_total?: number;
  side_sw?: number;
  side_ne?: number;
  setback_notes?: string;
}

export interface DevelopmentEnvelope {
  zoning_district: string;
  zoning_description: string;
  permitted_uses: string[];
  conditional_uses: string[];
  max_height_ft?: number;
  max_stories?: number;
  far?: number;
  max_lot_coverage_pct?: number;
  density_units_per_acre?: number;
  setbacks: Setbacks;
  buildable_area_sf?: number;
  parking_requirements?: string;
  lot_occupancy_pct?: number;
  height_source?: string;
  binding_constraint?: string;
}

export interface Scenario {
  name: string;
  unit_count_range?: string;
  description: string;
  constraints: string[];
  risk_level: RiskLevel;
  estimated_timeline: string;
  board_engagement: string;
}

export interface RiskMap {
  historic_overlay?: string;
  flood_zone?: string;
  building_category?: string;
  accommodation_overlay?: string;
  community_sensitivity?: string;
  recent_board_decisions?: string;
  environmental_constraints?: string;
}

export interface ProcessTimeline {
  required_boards: string[];
  review_sequence: string;
  estimated_meetings: string;
  estimated_months: string;
  current_backlog?: string;
  additional_permits: string[];
}

export interface CostFraming {
  permit_fees_estimate?: string;
  construction_cost_range?: string;
  tax_credit_eligibility?: string;
  bailey_bill_eligible?: boolean;
  total_cost_range?: string;
  construction_type?: string;
  base_hard_cost_range?: string;
  applicable_premiums?: string[];
  premium_adjusted_range?: string;
  all_in_estimate_range?: string;
  impact_fees_estimate?: string;
  cws_fee_warning?: string;
}

export interface AnalysisMetadata {
  generated_at: string;
  solver_version: string;
  ai_model: string;
  jurisdiction_module_version: string;
}

// --- Response ---

export interface AnalysisResponse {
  parcel: ParcelInfo;
  envelope: DevelopmentEnvelope;
  scenarios: [Scenario, Scenario, Scenario];
  risk_map: RiskMap;
  process_timeline: ProcessTimeline;
  cost_framing: CostFraming;
  metadata: AnalysisMetadata;
  confidence_tier: number; // 1, 2, or 3
  confidence_label: string;
  disclaimer?: string;
}
